### PR TITLE
Add warning log for direct path connectivity

### PR DESCRIPTION
**Pull request by anushka567**
### Description
 Using  the [storage.CheckDirectConnectivitySupported](https://pkg.go.dev/cloud.google.com/go/storage@v1.47.0#CheckDirectConnectivitySupported) function to check for direct connectivity support, this PR
adds a warning log message if direct connectivity is unavailable, including the bucket name and the error message.
This logging is only for GRPC data client.

Sample log:
```
Direct path connectivity unavailable for <bucket-name>, reason: storage: direct connectivity not detected
```

Time observed for figuring out direct connectivity :
For vm located in us-west1-b: 10159 msec(issue actively tracked in internal bug)
For vm located in us-central1-f : 365 msec
### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

File: internal/storage/storage_handle.go [hunk 1]
```diff
@@ -250,5 +253,12 @@ func (sh *storageClient) BucketHandle(bucketName string, billingProject string)
 		bucketName:    bucketName,
 		controlClient: sh.storageControlClient,
 	}
+
+	if sh.clientProtocol == "grpc" {
+		if err := storage.CheckDirectConnectivitySupported(ctx, bucketName, sh.storageClientOpts...); err != nil {
+			logger.Warnf("Direct connectivity unavailable for %s, error : %s", bucketName, err)
```
Comment (raj-prince) [inline internal/storage/storage_handle.go:259]: nit: remove space before `:`

File: internal/storage/storage_handle.go [hunk 2]
```diff
@@ -250,5 +253,12 @@ func (sh *storageClient) BucketHandle(bucketName string, billingProject string)
 		bucketName:    bucketName,
 		controlClient: sh.storageControlClient,
 	}
+
+	if sh.clientProtocol == "grpc" {
+		if err := storage.CheckDirectConnectivitySupported(ctx, bucketName, sh.storageClientOpts...); err != nil {
```
Comment (raj-prince) [inline internal/storage/storage_handle.go:258]: Do we know, how much time does it take? - Can we update the description with this information?

**Review by raj-prince — CHANGES_REQUESTED** [review]
Added few comments.

File: internal/storage/storage_handle.go [hunk 3]
```diff
@@ -204,7 +207,7 @@ func NewStorageHandle(ctx context.Context, clientConfig storageutil.StorageClien
 	// TODO: We will implement an additional check for the HTTP control client protocol once the Go SDK supports HTTP.
 	// TODO: Custom endpoints do not currently support gRPC. Remove this additional check once TPC(custom-endpoint) supports gRPC.
 	if clientConfig.EnableHNS && clientConfig.CustomEndpoint == "" {
-		clientOpts, err := createClientOptionForGRPCClient(&clientConfig)
+		clientOpts, err = createClientOptionForGRPCClient(&clientConfig)
```
[the remark this answers is not in the document]
Comment (anushka567) [reply internal/storage/storage_handle.go:226]: Updated. The clientOptions only depend on the clientconfig passed, so used createClientOptionForGRPCClient function at the start that can be later used for direct path detector initialization.

File: internal/storage/storage_handle.go [hunk 4]
```diff
@@ -50,12 +50,14 @@ type StorageHandle interface {
...
 	//
 	// A user-project is required for all operations on Requester Pays buckets.
-	BucketHandle(bucketName string, billingProject string) (bh *bucketHandle)
+	BucketHandle(context context.Context, bucketName string, billingProject string) (bh *bucketHandle)
 }
 
 type storageClient struct {
 	client               *storage.Client
 	storageControlClient *control.StorageControlClient
+	storageClientOpts    []option.ClientOption
```
[the remark this answers is not in the document]
Comment (anushka567) [reply internal/storage/storage_handle.go]: Updated.

File: internal/storage/storage_handle.go (hunk 1 above)
Comment (anushka567) [reply internal/storage/storage_handle.go]: Updated.

File: internal/storage/storage_handle.go [hunk 5]
```diff
@@ -50,12 +50,25 @@ type StorageHandle interface {
 	// to that project rather than to the bucket's owning project.
 	//
 	// A user-project is required for all operations on Requester Pays buckets.
-	BucketHandle(bucketName string, billingProject string) (bh *bucketHandle)
+	BucketHandle(context context.Context, bucketName string, billingProject string) (bh *bucketHandle)
```
Comment (raj-prince) [inline internal/storage/storage_handle.go:53]: nit: context -> ctx

File: internal/storage/storage_handle.go [hunk 6]
```diff
@@ -50,12 +50,25 @@ type StorageHandle interface {
...
 	client               *storage.Client
 	storageControlClient *control.StorageControlClient
+	directPathDetector   *gRPCDirectPathDetector
+}
+
+type gRPCDirectPathDetector struct {
+	clientOptions []option.ClientOption
+}
+
+func (pd *gRPCDirectPathDetector) isDirectPathPossible(ctx context.Context, bucketName string) bool {
```
Comment (raj-prince) [inline internal/storage/storage_handle.go:66]: As discussed offline, we will provide responsibility to client to log. This utility will only provide yes or not.

File: internal/storage/storage_handle_test.go [hunk 7]
```diff
@@ -51,7 +51,8 @@ func (testSuite *StorageHandleTest) TearDownTest() {
 
 func (testSuite *StorageHandleTest) TestBucketHandleWhenBucketExistsWithEmptyBillingProject() {
 	storageHandle := testSuite.fakeStorage.CreateStorageHandle()
-	bucketHandle := storageHandle.BucketHandle(TestBucketName, "")
+	ctx := context.Background()
```
Comment (gargnitingoogle) [inline internal/storage/storage_handle_test.go:54]: nit: Code simplification.

Add this to SetupTest() 

```go
testSuite.ctx = context.Background()
```

And then just use `testSuite.ctx` directly in each test function.

**Review by gargnitingoogle — APPROVED** [review]
Minor comments. LGTM .

File: internal/storage/storage_handle.go [hunk 8]
```diff
@@ -250,5 +268,10 @@ func (sh *storageClient) BucketHandle(bucketName string, billingProject string)
 		bucketName:    bucketName,
 		controlClient: sh.storageControlClient,
 	}
+	if sh.directPathDetector != nil {
+		if err := sh.directPathDetector.isDirectPathPossible(ctx, bucketName); err != nil {
+			logger.Warnf("Direct path connectivity unavailable for %s, reason: %s", bucketName, err)
```
[the remark this answers is not in the document]
Comment (anushka567) [reply internal/storage/storage_handle.go]: Going with suggestion#2

File: internal/storage/storage_handle_test.go (hunk 7 above)
Comment (anushka567) [reply internal/storage/storage_handle_test.go]: Updated.

**Review by raj-prince — APPROVED** [review]
LGTM.

Could you please update the description with the logs, we get in case direct-path is not possible?

**Pull request merged**